snapshots/overlay/overlayutils/check.go

```diff
@@ -167,3 +169,100 @@ func NeedsUserXAttr(d string) (bool, error) {
 	}
 	return true, nil
 }
+
+const (
+	MetacopyDefault     = 0
+	MetacopySupported   = 1
+	MetacopyUnsupported = 2
+	MetacopyCheckFailed = 3
+)
+
+func SupportsMetacopy(root string) (int, error) {
+	metacopyState := MetacopyCheckFailed
+
+	if _, err := os.Stat("/sys/module/overlay/parameters/metacopy"); err == nil {
+		metacopy, err := os.ReadFile("/sys/module/overlay/parameters/metacopy")
+		if err == nil {
+			if metacopy[0] == 'Y' {
+				metacopyState = MetacopyDefault
+			} else {
+				metacopyState = MetacopySupported
+			}
+		} else {
+			log.L.WithError(err).Warnf("Unable to check metacopy state using sysfs")
+		}
+	}
+
+	testData := "Let me use metacopy!"
+
+	testDir, err := os.MkdirTemp(root, "metacopy_test")
```

Comment:
Just curious, in which case can we fail to check via the sysfs files and we want to do a test mount to see if it is available? When sysfs is not mounted? Is there some other case?

I'm ignorant here, but do we care about sysfs not being mounted? Maybe it is common in some scenarios that I don't know about?

Reply:
Yes, this is for the cases when there is no sysfs mounted or there are some specific LSM policies. But I also not sure that this is the case. So, maybe I will simplify this check to use sysfs only and move this to integration test.

Reply:
I don't have a strong opinion here, not sure if in other parts of the code we already rely on sysfs maybe? I'd let others weight in.

Reply:
Yes, in this case it will work like this:
check via sysfs is ok? -> return, otherwise -> slowpath, check via xattrs

Reply:
> I don't have a strong opinion here, not sure if in other parts of the code we already rely on sysfs maybe? I'd let others weight in. Maybe @dmcgowan that also reviewed this?
Another argument for simplification is the next:
overlay snapshotter already use sysfs to check overlay "index=off" option.
https://github.com/containerd/containerd/blob/681aaf68b7dcbe08a51c3372cbb8f813fb4466e0/snapshots/overlay/overlay.go#L569

Reply:
@artqzn right, I think that is what I was thinking about that we already use it. My vote is to remove all of the rest, then.